[Merge] 学生操作功能设计

fanfan187 opened the pull request:
---

## 变更内容

- **新增/修改/删除**：
    - 新增 `StudentController`,  `StudentLogic`, `StudentDisableDTO`, `StudentVO`,  `StudentService`
    - 对 `StringConstant`, `StudentDAO`, `UserDAO`, `StudentDTO`进行修改以实现接口功能
    - 补充测试类

---

## 测试内容

- 单元测试：DAO层已完成
- 接口测试：已完成

---

fanfan187 committed:
feat(models): 添加学生视图对象

- 新增 StudentVO 类，用于表示学生信息
- 添加了学号、姓名、性别、学院、专业、班级等字段
- 使用了数据校验注解，确保数据的完整性

fanfan187 committed:
feat(StudentController): 添加学生管理功能

- 新增 StudentController 类，实现学生管理相关接口
-包含学生信息查看、学生列表查询、添加学生、停用学生、删除学生和编辑学生等功能

fanfan187 committed:
feat(student): 添加学生信息相关功能

- 新增学生信息查看接口
- 实现学生信息添加功能
- 增加参数校验逻辑
- 定义学生信息相关的正则表达式常量

fanfan187 committed:
Merge remote-tracking branch 'origin/develop' into feature/student-operate

fanfan187 committed:
Merge remote-tracking branch 'origin/develop' into feature/student-operate

fanfan187 committed:
Merge remote-tracking branch 'origin/develop' into feature/student-operate

# Conflicts:
#	src/main/java/com/frontleaves/scheduling/constants/StringConstant.java

fanfan187 committed:
refactor(scheduling): 重构学生相关字段并更新验证注解
- 将性别字段从 Integer 类型改为 Boolean 类型
- 添加 gradeUuid 和 userUuid 字段- 更新字段验证注解，增加不能为空的验证
- 移除多余的 Min 和 Max 注解

fanfan187 committed:
feat(student): 实现学生信息管理功能

- 添加学生信息列表查询接口，支持分页和多条件筛选
- 实现学生信息添加、编辑和停用功能
- 新增学生相关数据传输对象和视图对象
- 优化学生信息的数据库查询和更新逻辑

fanfan187 committed:
feat(student): 实现了删除学生的逻辑

- 注释掉禁用学生的相关代码
- 实现删除学生功能
- 更新学生逻辑层，增加学生存在性检查
- 调整学生数据对象，移除不必要的字段
- 优化学生视图对象转换逻辑

fanfan187 committed:
feat(student): 启用或禁用学生账户功能

- 新增 disableStudent 方法，根据学生 UUID 和禁用标志更新学生账户状态- 在 StudentDTO 中添加 status 字段表示学生状态
- 更新 StudentController 中的 disableStudent 方法，支持启用或禁用学生账户
- 优化 StudentDAO 和 StudentLogic 中的相关方法，支持学生状态更新

fanfan187 committed:
test(student): 补充学生相关测试用例并优化学生模型

- 在 StudentTest 类中添加测试学生列表和编辑学生信息的方法
- 优化 StudentDO、StudentDTO 和 StudentVO 中的字段命名
- 更新 StudentVO 类，使用 Lombok 注解简化代码

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/StudentController.java:
```diff
@@ -0,0 +1,184 @@
+     */
+    @PostMapping("")
+    public @NotNull ResponseEntity<BaseResponse<StudentDTO>> addStudent(
+            @RequestBody StudentVO studentVO
```
数据验证

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/StudentController.java:
```diff
@@ -0,0 +1,184 @@
+
+        // VO -> DO
+        StudentDO studentDO = new StudentDO();
+        BeanUtils.copyProperties(studentVO, studentDO);
```
【不算问题】DO 尽量不出现在 Controller。下方的 editStudent 你写法也是直接数据 VO 到 DO，在这里怎么转换了

fanfan187 replied:
因为上一个接口MajorController里面也是这样写的，就以为这种是正确的了😂

xiaolfeng replied:
之前的就没让你改了，实际上不是很推荐。主要是 DO 本身是对数据库进行操作，确实不该出现在控制层。控制层主要处理数据验证，Logic 走向（如何进入服务层）。

并且第二点，如果你提前转为 DO，那么 VO 和 DO 大部分情况下 VO 数据量较少。那么转为 DO 会有很多数据是空数据，你还需要额外花精力要对数据是否为空或者进一步逻辑操作要进一步考虑。VO 可以直接拿到数据对数据进行处理，处理完成后一起合并为 DO

xiaolfeng replied:
尤其是 VO 部分数据是可填写的情况下，转为 DO，有些位置原本是空的，和部分选填位置是空的，就会造成数据验证麻烦，要自己去验证数据一个个去比对

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/daos/StudentDAO.java:
```diff
@@ -229,15 +233,102 @@ public StudentDO getStudentByUserUuid(String userUuid) {
+        // 选填字段均为空,返回空页
+        if (CharSequenceUtil.isBlank(clazz) && CharSequenceUtil.isBlank(name) && CharSequenceUtil.isBlank(id)) {
+            return new Page<>(page, size);
+        }
```
选填为空代表查询条件为空，没有查询条件，不是不需要查询内容

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/daos/StudentDAO.java:
```diff
@@ -229,15 +233,102 @@ public StudentDO getStudentByUserUuid(String userUuid) {
+     * @param id 学生学号，可为空
+     * @return 返回包含学生信息的页面对象
+     */
+    public Page<StudentDO> listStudents(int page, int size, Boolean isDesc,
```
缺缓存

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,241 @@
+        if (studentDO == null) {
+            throw new BusinessException("学生不存在", ErrorCode.NOT_EXIST);
+        }
+        return convertToStudentDTO(studentDO);
```
可以直接使用 BeanUtil.toBean 由于你的 DTO 和 DO 没有多余的数据（多出 status），而且在你的 this.convertToStudentDTO 也不存在 status 的数据处理，所以该部分可以变成 BeanUtil.toBean

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,241 @@
+        }).toList();
+
+        // 构造并返回PageDTO对象
+        return new PageDTO<>(dtoList, resultPage.getTotal(), resultPage.getSize(), resultPage.getCurrent());
```
在 ProjectUtil 里含一个 pageToPageDTO 的方法，可以直接转换

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,241 @@
+    @Override
+    public StudentDO addStudent(StudentDO studentVO) {
+        // 校检数据完整性
+        if (studentVO == null || studentVO.getId() == null || studentVO.getName() == null || studentVO.getGender() == null || studentVO.getGradeUuid() == null || studentVO.getClazz() == null) {
```
数据校验不应该出现在 Logic，对应于 controller 部分，在 Controller 部分缺少注解，可直接注解判断

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,241 @@
+
+        // 存入数据库
+        studentDAO.save(studentDO);
+        convertToStudentDTO(studentDO);
```
新建的 DO 直接转成 DTO 数据不完整，要从数据库重新提取数据（例如：缺少用户主键、创建时间和删除时间「还有部分数据库自动填写的数据」）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/models/vo/StudentVO.java:
```diff
@@ -0,0 +1,47 @@
+@Accessors(chain = true)
+public class StudentVO {
+    @NotBlank(message = "学生UUID不能为空")
+    private String studentUuid;
```
创建用户不应该出现 UUID，创建用户的 UUID 应该交给系统自行创建

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,241 @@
+        }
+
+        StudentDO studentDO = new StudentDO();
+        studentDO.setStudentUuid(studentVO.getStudentUuid())
```
没有对外键约束做判断，如果数据传入错误或格式错误均会报未定义过的错

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,241 @@
+        }
+
+        // 将视图对象中的属性复制到实体对象中,以准备更新数据库
+        BeanUtils.copyProperties(studentVO, studentDO);
```
没有对外键相关内容进行验证

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,241 @@
+     * @param studentDO 学生数据对象,包含从数据库中获取的学生信息
+     * @return 返回一个包含相同信息的学生数据传输对象
+     */
+    private StudentDTO convertToStudentDTO(StudentDO studentDO) {
```
可以使用 BeanUtil 替换掉

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/models/dto/StudentDTO.java:
```diff
@@ -77,4 +82,9 @@ public class StudentDTO {
+    /**
+     * 学生状态
+     */
+    public Boolean status;
```
public？

xiaolfeng reviewed (changes requested):
1. 编码不规范
2. 没有对数据进行验证
3. 编码格式不统一

xiaolfeng commented:
另外，对于创建学生和编辑学生里，只需要确认学生的班级。就可以定位到 grade_uuid, department, major 所以对于 StudentVO 来说，是不需要额外让用户输入  grade_uuid, department, major。
并且 UserUuid 应该是不存在的，这个 UUID 应该交给系统登录时候自动创建，而并非手动或在 add student 或 edit student 的时候添加

xiaolfeng commented:
![image](https://github.com/user-attachments/assets/81ebc155-127f-4b43-b1ea-a112f74a11f5)
回复你这个，我没找到，创建学生的时候，账号是用户自行创建的，而不是管理员手动创建的（在设计思路上已经避免了外键多个相连，所以管理员不设置允许操作 user_uuid 直接避免操作外键）

添加学生就是添加学生，不包含创建用户信息，用户信息由用户登录后自动匹配上绑定。

另外你的截图给的是 student_uuid 这个 uuid 应该交给系统生成，系统主键是指向数据（原则上不应该让用户知道），对于其他外键约束可以通过其他接口通过前端对接。但是创建数据来说，主键一定是新建一个，而不是引用其他的，就算你要填写，也是交给前端实现创建 uuid（对于前端来说，前端尽量避免数据逻辑操作，最好是获取数据或对接数据直接进行操作，不需要太多数据操作）

不符合前端的设计原则

fanfan187 committed:
feat(student): 优化学生信息处理逻辑

- 新增 `ClassMappingDTO` 和 `MappingUtil` 用于班级映射
- 重构 `StudentController`、`StudentDAO`、`StudentLogic 和 `StudentService`
- 优化学生信息添加和编辑流程，自动处理年级、学院和专业信息
- 增加 Redis 缓存支持，提高学生信息查询效率
- 更新相关测试用例

fanfan187 committed:
Merge remote-tracking branch 'origin/develop' into feature/student-operate

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/StudentController.java:
```diff
@@ -0,0 +1,173 @@
+
+        // 检查编辑结果,如果返回null,则表示学生不存在
+        if (studentDTO == null) {
+            return ResultUtil.error(ErrorCode.NOT_EXIST, "学生不存在", null);
```
有错误请使用 throw，不要使用 ResultUtil.error

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/daos/StudentDAO.java:
```diff
@@ -229,15 +239,143 @@ public StudentDO getStudentByUserUuid(String userUuid) {
+            // 查看学生是否毕业
+            if (isGraduated != null) {
+                queryWrapper.eq(StudentDO::getGraduated, isGraduated)
+                        .or().isNull(StudentDO::getGraduated);
```
is_graduated 在数据库是非空字段, isNull 在这里是多余的

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/daos/StudentDAO.java:
```diff
@@ -229,15 +239,143 @@ public StudentDO getStudentByUserUuid(String userUuid) {
+                        .collect(Collectors.toList());
+                Page<StudentDO> cachePage = new Page<>();
+                cachePage.setRecords(studentList);
+                transaction.commit();
```
这一段数据解析是有问题的，在 ProjectUtil 有现成方法

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/utils/MappingUtil.java:
```diff
@@ -0,0 +1,56 @@
+ * @since v1.0.0
+ */
+@Component
+public class MappingUtil {
```
工具类尽量不使用注入（侵入式），尽量用 static。否则多处重复调用容易造成 Bean 反复注入

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/daos/StudentDAO.java:
```diff
@@ -229,15 +239,143 @@ public StudentDO getStudentByUserUuid(String userUuid) {
+                    .setMajor(classMapping.getMajorUuid());
+
+            // 更新数据库
+            boolean success = this.lambdaUpdate()
```
完全独立，那你获取 studentDO 除了检查数据是否存在，完全没有用。你的 BeanUtil 也是没有用到

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,216 @@
+    ) {
+        // 调用DAO层方法获取分页学生数据
+        Page<StudentDO> resultPage = studentDAO.listStudents(page, size, isDesc, clazz, isGraduated, name, id);
+
```
在 DAO 里面，如果找不到数据应该返回 null（而并非存在该实体），所以逻辑层要起到保险作用要判空，并且判空后保证返回用户端数据正常应该不可以返回 null

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,216 @@
+        String studentUuid = UuidUtil.generateUuidNoDash();
+
+        // 校验学生姓名
+        if (!Pattern.matches(StringConstant.Regular.STUDENT_NAME_REGULAR_EXPRESSION, studentDTO.getName())) {
```
你在 VO 已经对数据进行校验非空了，那么正则验证为什么不放进去呢

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/StudentLogic.java:
```diff
@@ -0,0 +1,216 @@
+            throw new BusinessException("学生姓名格式错误", ErrorCode.PARAMETER_INVALID);
+        }
+        // 检查班级是否为空(进行外键映射)
+        if (studentDTO.getClazz() == null || studentDTO.getClazz().isBlank()) {
```
这个也是，并且名字有正则验证，为什么作为班级（有外键约束的 UUID 反而不验证呢）

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/utils/MappingUtil.java:
```diff
@@ -0,0 +1,56 @@
+     * @return ClassMappingDTO 包含班级映射信息的对象,包括年级UUID、院系UUID和专业UUID
+     * @throws BusinessException 当未找到对应班级信息时抛出
+     */
+    public ClassMappingDTO getClassMappingByClazz(String clazz) {
```
该工具方法为什么不直接在对应 DAO 写呢

fanfan187 replied:
我之前想的是这样可以做到各层各司其职并且看起来很清晰，想到了经常用的常量会放到SringConstant里那样，也类似有时候一个文件里有重复代码，又新增了一个方法来处理它的这种感觉

xiaolfeng replied:
实际上操作属于 DAO 的内容，DAO 也是可以抽象出来的。工具类目的是处理一个过程。数据操作数据交流交给 DAO

> 各层各司其职并且看起来很清晰

代码里也是操作数据库，所以没问题。如果你想更各司其职分布操作的话。那么这个 DAO 应该返回 DO，你转换的 MappingDTO 应该在 Logic 完成转换。然后你的另外一个地方调用这个 Logic

fanfan187 committed:
refactor(student): 解决冲突

xiaolfeng reviewed (changes requested):
规约和规范

fanfan187 committed:
refactor(student): 解决冲突2.0

- 对文件进行了冲突解决，并将搁置的文件取消托管，统一提交
